### ISSUE TITLE
Migrate to C++11

### ISSUE THREAD

**Issue by hobu** [libspatialindex/libspatialindex#60]
The codebase can be migrated to C++11, which will allow us to remove a number of homegrown tools/utilities that didn't really exist when libspatialindex was first developed. Some of the things we will attempt to refactor:

* remove Tools::SharedPtr in exchange for std::shared_ptr
* remove threading altogether
* Implement gtest for test runs
* ...

**Comment by davisp** [comment]
I'd also like to remove PointerPool and PoolPointer. Though I think that may involve a bit of code churn so it might be best as its own PR.

I'd also like to remove all references to the delete operator as well for memory management. A lot of this means replacing our member variable raw pointers with std::unique_ptr. There are some other buffers that only exist for the scope of a single function that we could just use standard dynamic stack allocation with as well.

**Comment by hobu** [comment]
Yes, make a PR against the cpp-11 branch. Let's leave master alone until we're ready to bless it as ready and merge it.

**Comment by citibeth** [comment]
This should also help eliminate manual memory management, which is a source of possible errors and memory leaks.  `new()`, `delete()` and `memcpy()` should not be necessary with C++11.  See #54

**Comment by citibeth** [comment]
You can get rid of things like `IVisitor` and use `std::function` instead.  MUCH easier.

**Comment by r-barnes** [comment]
Just wondering where things are at on this? I'm trying to include libspatialindex directly in a C++ library and finding it interesting going.

**Comment by hobu** [comment]
> Just wondering where things are at on this? I'm trying to include libspatialindex directly in a C++ library and finding it interesting going.

There hasn't been much contribution lately. If you are able to make a go at working this topic, we can look to merge things.

**Comment by r-barnes** [comment]
Thanks! Honestly, it's a lot to dig into and familiarize one's self with. I don't think I have the time, and I'm not in a position to fund work. I don't expect you're in a better position.

I wonder, though, if converting the library into thread-safe C++11 could be cast as a project for [Google Summer of Code](https://developers.google.com/open-source/gsoc/), perhaps with the [OGC](http://www.opengeospatial.org/) or another group serving as the coordinating organization? I'm not super-familiar with the process, but I think GDAL's had some success with this in the past. The downside is that this still requires time spent mentoring.

**Comment by hobu** [comment]
It's probably too big of a project for a GSoC effort unless the person was already well-versed in C++11. The biggest impediment is there's no testing framework for the library that's continuous integration-friendly, so it is difficult to develop things in a confident fashion.

It's probably two or three GSoC's of effort:

1) Continue @normanb's efforts to regularize the tests
2) Replace all of the `spatialindex::tools` stuff with C++11
3) Attack the library itself with C++11 improvements.
